### PR TITLE
HPCC-26525 Allow stub resources to be configurable.

### PR DESCRIPTION
**Pull request by jakesmith**
And introduce template helpers to convert between K8s memory
and cpu strings

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: helm/hpcc/values.schema.json [hunk 1]
```diff
@@ -1650,6 +1658,23 @@
...
+      "type": "object",
+      "description" : "resource definitions per instance for stub components",
+      "properties": {
+        "memory": {
+          "description": "default per stub instance memory requirements",
+          "type": "string",
+          "default": "200Mi"
+        },
+        "cpu": {
+          "description": "default per stub instance micro cpu requirements",
```
Comment (richardkchapman) [inline helm/hpcc/values.schema.json:1672]: Do you mean milli rather than micro?

File: helm/hpcc/values.schema.json [hunk 2]
```diff
@@ -237,6 +237,14 @@
         "noResourceValidation": {
           "type": "boolean"
         },
+        "stubInstanceResources": {
+          "$ref": "#/definitions/stubInstanceResources"
+        },
+        "stubPerInstanceCPU": {
+          "description": "default per stub instance micro cpu requirements",
```
Comment (richardkchapman) [inline helm/hpcc/values.schema.json:244]: milli rather than micro I think?

File: helm/hpcc/templates/_helpers.tpl [hunk 3]
```diff
@@ -1548,3 +1554,82 @@ A template to output a merged environment. Pass in a list with global then local
...
+ {{- printf "%d" (int (mulf (float64 .value) 1000.0)) -}}
+{{- end -}}
+{{- end -}}
+
+{{/*
+A template to convert bytes into a k8s human friendly string
+Pass in value
+*/}}
+{{- define "hpcc.bytesToK8sMemoryString" -}}
+{{- /* NB: Rounds down to units that are a 10th of unit that values is larger than */ -}}
```
Comment (richardkchapman) [inline helm/hpcc/templates/_helpers.tpl:1621]: Does it? How? 1000th maybe...

File: helm/hpcc/templates/_helpers.tpl [hunk 4]
```diff
@@ -860,13 +860,19 @@ resources:
...
+{{- $memPerInstance := $stubInstanceResources.memory | default "200Mi" -}}
+{{- $milliCPUs := int (include "hpcc.k8sCPUStringToMilliCPU" (dict "value" $milliCPUPerInstance)) -}}
+{{- $bytes := int64 (include "hpcc.k8sMemoryStringToBytes" (dict "value" $memPerInstance)) -}}
+{{- $totalBytes := mul .instances $bytes }}
 resources:
   limits:
-    cpu: "50m"
-    memory: {{ (printf "%dM" (mul .instances 100)) | quote }}
+    cpu: {{ printf "%dm" (mul .instances $milliCPUs) | quote }}
+    memory: {{ include "hpcc.bytesToK8sMemoryString" (dict "value" $totalBytes) | quote }}
```
Comment (richardkchapman) [inline helm/hpcc/templates/_helpers.tpl:875]: If you are just passing a single value, there's not really a need to use a dict and pass it by name...

File: helm/hpcc/values.schema.json [hunk 5]
```diff
@@ -237,6 +237,14 @@
         "noResourceValidation": {
           "type": "boolean"
         },
+        "stubInstanceResources": {
+          "$ref": "#/definitions/stubInstanceResources"
+        },
+        "stubPerInstanceCPU": {
```
Comment (richardkchapman) [inline helm/hpcc/values.schema.json:243]: Is this used? If so how does it differ from stubInstanceResources.cpu ?

**Review by richardkchapman — CHANGES_REQUESTED** [review]
Some minor questions

File: helm/hpcc/templates/_helpers.tpl (hunk 3 above)
Comment (jakesmith) [reply helm/hpcc/templates/_helpers.tpl:1621]: yes, meant a 1000th will change.
e.g. 2Gi => 2000Mi, 2500Mi => 2500Mi.
It was so I didn't have to deal with fractional cases

File: helm/hpcc/values.schema.json (hunk 5 above)
Comment (jakesmith) [reply helm/hpcc/values.schema.json:243]: no, was left in by mistake, will remove.

File: helm/hpcc/values.schema.json (hunk 1 above)
Comment (jakesmith) [reply helm/hpcc/values.schema.json:1672]: yes mean milli, will correct

File: helm/hpcc/templates/_helpers.tpl (hunk 4 above)
Comment (jakesmith) [reply helm/hpcc/templates/_helpers.tpl:875]: I didn't know you could, thought had to be a scope, which meant a dictionary..

**Review by richardkchapman — APPROVED** [review]

**Comment by jakesmith** [comment]
@richardkchapman - squashed

**Pull request merged**